pkgs/development/compilers/openjdk/generic.nix

```diff
@@ -0,0 +1,664 @@
+          ./11/patches/increase-javadoc-heap.patch
+      )
+    ]
+    ++ lib.optionals atLeast17 [
+      (
+        if atLeast21 then
+          ./21/patches/ignore-LegalNoticeFilePlugin-jdk18.patch
+        else
+          ./17/patches/ignore-LegalNoticeFilePlugin-jdk17.patch
+      )
+    ]
+    ++ lib.optionals (!atLeast21) [
+      (
+        if atLeast17 then
+          ./17/patches/fix-library-path-jdk17.patch
+        else if atLeast11 then
+          ./11/patches/fix-library-path-jdk11.patch
+        else
+          ./8/patches/fix-library-path-jdk8.patch
+      )
+    ]
+    ++ lib.optionals (atLeast17 && !atLeast23) [
+      # -Wformat etc. are stricter in newer gccs, per
+      # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79677
+      # so grab the work-around from
+      # https://src.fedoraproject.org/rpms/java-openjdk/pull-request/24
+      (fetchurl {
+        url = "https://src.fedoraproject.org/rpms/java-openjdk/raw/06c001c7d87f2e9fe4fedeef2d993bcd5d7afa2a/f/rh1673833-remove_removal_of_wformat_during_test_compilation.patch";
+        sha256 = "082lmc30x64x583vqq00c8y0wqih3y4r0mp1c4bqq36l22qv6b6r";
+      })
+    ]
+    ++ lib.optionals (featureVersion == "17") [
+      # Patch borrowed from Alpine to fix build errors with musl libc and recent gcc.
+      # This is applied anywhere to prevent patchrot.
+      (fetchurl {
+        url = "https://git.alpinelinux.org/aports/plain/community/openjdk17/FixNullPtrCast.patch?id=41e78a067953e0b13d062d632bae6c4f8028d91c";
+        sha256 = "sha256-LzmSew51+DyqqGyyMw2fbXeBluCiCYsS1nCjt9hX6zo=";
+      })
+    ]
+    ++ lib.optionals atLeast11 [
+      # Fix build for gnumake-4.4.1:
+      #   https://github.com/openjdk/jdk/pull/12992
+      (fetchpatch {
+        name = "gnumake-4.4.1";
+        url = "https://github.com/openjdk/jdk/commit/9341d135b855cc208d48e47d30cd90aafa354c36.patch";
+        hash = "sha256-Qcm3ZmGCOYLZcskNjj7DYR85R4v07vYvvavrVOYL8vg=";
+      })
+    ]
+    ++ lib.optionals (featureVersion == "17") [
+      # Backport fixes for musl 1.2.4 which are already applied in jdk21+
+      # Fetching patch from chimera because they already went through the effort of rebasing it onto jdk17
+      (fetchurl {
+        name = "lfs64.patch";
+        url = "https://raw.githubusercontent.com/chimera-linux/cports/4614075d19e9c9636f3f7e476687247f63330a35/contrib/openjdk17/patches/lfs64.patch";
+        hash = "sha256-t2mRbdEiumBAbIAC0zsJNwCn59WYWHsnRtuOSL6bWB4=";
+      })
+    ]
+    ++ lib.optionals (!headless && enableGtk) [
+      (
+        if atLeast17 then
+          ./17/patches/swing-use-gtk-jdk13.patch
+        else if atLeast11 then
+          ./11/patches/swing-use-gtk-jdk10.patch
+        else
+          ./8/patches/swing-use-gtk-jdk8.patch
+      )
+    ];
+
+  nativeBuildInputs =
+    [
+      pkg-config
+    ]
+    ++ lib.optionals atLeast11 [
+      autoconf
+    ]
+    ++ lib.optionals (!atLeast11) [
+      lndir
+    ]
+    ++ [
+      unzip
+    ]
```

Comment:
No, the ordering matters for the derivation hash. This is not meant to be a clean‐up but rather an accurate 0‐rebuild refactoring of the existing derivation, so that we can be sure the refactoring is correct. I will clean it up separately.

Reply:
I have less faith in Nix today than I did yesterday. Input packages are seriously not sorted and deduplicated?? Gah.

Reply:
I think this is the difference of using a list instead of a set (not that Nix has a native set data type), and like @emilazy said this actually can cause a build to differ (e.g.: this will change the order of applications appears in `$PATH` for `nativeBuildInputs`), so I don't think this is unresonable behavior.